From d3ff444e99f12279b86dd651055515344763dab1 Mon Sep 17 00:00:00 2001
From: t <tristan75015@gmail.com>
Date: Tue, 13 Oct 2026 21:36:05 +0200
Subject: [PATCH] MTest checks the numerical type of the generic material
 properties

MTest only checked the numerical type of the generic behaviours: a generic
material property or a generic evolution generated with another type than
double (real_type option of mfront) was read as double, giving wrong values
without error. ExternalLibraryManager::checkRealSize gives the check and
its message, used by the three loaders of MTest.
---
 include/TFEL/System/ExternalLibraryManager.hxx | 11 +++++++++++
 mtest/src/GenericBehaviour.cxx                 | 10 +---------
 mtest/src/GenericEvolution.cxx                 |  2 ++
 mtest/src/GenericMaterialProperty.cxx          |  2 ++
 src/System/ExternalLibraryManager.cxx          | 13 +++++++++++++
 5 files changed, 29 insertions(+), 9 deletions(-)

diff --git a/include/TFEL/System/ExternalLibraryManager.hxx b/include/TFEL/System/ExternalLibraryManager.hxx
index cc096f14d..e3028bffb 100644
--- a/include/TFEL/System/ExternalLibraryManager.hxx
+++ b/include/TFEL/System/ExternalLibraryManager.hxx
@@ -89,6 +89,17 @@ namespace tfel::system {
      * material property)
      */
     unsigned short getRealSize(const std::string&, const std::string&);
+    /*!
+     * \brief check that the numerical type of an entry point generated by
+     * the `generic` or `generic-parallel` interfaces has the expected size
+     * (see `getRealSize`), and throw an exception otherwise
+     * \param[in] l: library name
+     * \param[in] f: entry point name
+     * \param[in] s: expected size in bytes (`sizeof(double)` for MTest)
+     */
+    void checkRealSize(const std::string&,
+                       const std::string&,
+                       const unsigned short);
     /*!
      * \param[in] l : name of the library
      * \param[in] f : name of function or mechanical behaviour
diff --git a/mtest/src/GenericBehaviour.cxx b/mtest/src/GenericBehaviour.cxx
index 975c9d2d2..f8896d8d2 100644
--- a/mtest/src/GenericBehaviour.cxx
+++ b/mtest/src/GenericBehaviour.cxx
@@ -140,15 +140,7 @@ namespace mtest {
     auto& elm = ExternalLibraryManager::getExternalLibraryManager();
     // MTest computes in double: a behaviour generated with another
     // numerical type (real_type option of mfront) would be read wrongly
-    const auto rs = elm.getRealSize(l, b);
-    tfel::raise_if(rs != sizeof(real),
-                   "GenericBehaviour::GenericBehaviour: behaviour '" + b +
-                       "' in library '" + l + "' uses a numerical type of " +
-                       std::to_string(rs) + " bytes, whereas MTest requires " +
-                       std::to_string(sizeof(real)) +
-                       " bytes (double): the library must be generated "
-                       "without the real_type option of mfront, or with "
-                       "real_type set to double");
+    elm.checkRealSize(l, b, sizeof(real));
     const auto f = b + "_" + ModellingHypothesis::toString(h);
     this->fct = elm.getGenericBehaviourFunction(l, f);
     if (this->stype == 1u) {
diff --git a/mtest/src/GenericEvolution.cxx b/mtest/src/GenericEvolution.cxx
index 6581f21ca..249aab470 100644
--- a/mtest/src/GenericEvolution.cxx
+++ b/mtest/src/GenericEvolution.cxx
@@ -25,6 +25,8 @@ namespace mtest {
     using namespace tfel::system;
     using ELM = tfel::system::ExternalLibraryManager;
     auto& elm = ELM::getExternalLibraryManager();
+    // MTest computes in double (real_type option of mfront)
+    elm.checkRealSize(l, fn, sizeof(real));
     this->f = elm.getGenericMaterialProperty(l, fn);
     if (elm.getGenericMaterialPropertyNumberOfVariables(l, fn) != 0u) {
       elm.getGenericMaterialPropertyVariables(this->vnames, l, fn);
diff --git a/mtest/src/GenericMaterialProperty.cxx b/mtest/src/GenericMaterialProperty.cxx
index d511d1559..32089deb8 100644
--- a/mtest/src/GenericMaterialProperty.cxx
+++ b/mtest/src/GenericMaterialProperty.cxx
@@ -36,6 +36,8 @@ namespace mtest {
           "' has not been generated using the `Generic` interface");
     }
     auto& elm = ELM::getExternalLibraryManager();
+    // MTest computes in double (real_type option of mfront)
+    elm.checkRealSize(l, f, sizeof(real));
     this->fct = elm.getGenericMaterialProperty(l, f);
   }  // end of GenericMaterialProperty
 
diff --git a/src/System/ExternalLibraryManager.cxx b/src/System/ExternalLibraryManager.cxx
index ed68cf0f4..4bfefd480 100644
--- a/src/System/ExternalLibraryManager.cxx
+++ b/src/System/ExternalLibraryManager.cxx
@@ -337,6 +337,19 @@ namespace tfel::system {
     return static_cast<unsigned short>(n);
   }  // end of getRealSize
 
+  void ExternalLibraryManager::checkRealSize(const std::string& l,
+                                             const std::string& f,
+                                             const unsigned short s) {
+    const auto rs = this->getRealSize(l, f);
+    raise_if(rs != s, "ExternalLibraryManager::checkRealSize: entry point '" +
+                          f + "' in library '" + l +
+                          "' uses a numerical type of " + std::to_string(rs) +
+                          " bytes, whereas " + std::to_string(s) +
+                          " bytes are expected: the library must be "
+                          "generated with the expected type (real_type "
+                          "option of mfront)");
+  }  // end of checkRealSize
+
   bool ExternalLibraryManager::contains(const std::string& l,
                                         const std::string& s) {
     const auto lib = this->loadLibrary(l);
-- 
2.43.0

